katdal/h5datav2.py

```diff
@@ -443,6 +453,66 @@ def __str__(self):
         return '\n'.join(descr)
 
     @property
+    def _weights_keep(self):
+        known_weights = [row[0] for row in getattr(self, '_weights_description', [])]
+        return [known_weights[ind] for ind in self._weights_select]
+
+    @_weights_keep.setter
+    def _weights_keep(self, names):
+        known_weights = [row[0] for row in getattr(self, '_weights_description', [])]
+        # Ensure a sequence of weight names
+        names = known_weights if names == 'all' else \
+            names.split(',') if isinstance(names, basestring) else names
+        # Create index list for desired weights
+        selection = []
+        for name in names:
+            try:
+                selection.append(known_weights.index(name))
+            except ValueError:
+                logger.warning("%r is not a legitimate weight type for this file, "
+                               "supported ones are %s" % (name, known_weights))
+        if known_weights and not selection:
+            logger.warning('No valid weights were selected - setting all weights to 1.0 by default')
+        self._weights_select = selection
+
+    @property
+    def _flags_keep(self):
+        if not hasattr(self, '_flags_description'):
+            return []
```

Comment:
I'm guessing these various `hasattr` tests are because the base class touches `_flags_keep` and `_weights_keep` before the derived class constructor has completed. That has a bit of code smell about it, and at a minimum needs some comments. It might be cleaner (if a bit less efficient) to just compute the selection when accessing the `weights` property rather than when `_weights_keep` is set.